src/main/java/hudson/tasks/junit/CaseResult.java

```diff
@@ -244,14 +251,17 @@ static String fixNULs(String stdio) { // JENKINS-71139
     }
 
     /**
-     * Flavor of {@link #possiblyTrimStdio(Collection, boolean, String)} that doesn't try to read the whole thing into memory.
+     * Flavor of {@link #possiblyTrimStdio(Collection, StdioRetention, String)} that doesn't try to read the whole thing into memory.
      */
     @SuppressFBWarnings(value = "DM_DEFAULT_ENCODING", justification = "Expected behavior")
-    static String possiblyTrimStdio(Collection<CaseResult> results, boolean keepLongStdio, File stdio) throws IOException {
+    static String possiblyTrimStdio(Collection<CaseResult> results, StdioRetention stdioRetention, File stdio) throws IOException {
         long len = stdio.length();
-        if (keepLongStdio && len < 1024 * 1024) {
+        boolean keepAll = stdioRetention == StdioRetention.all
+                || (stdioRetention == StdioRetention.failed && hasFailures(results));
+        if (keepAll && len < 1024 * 1024) {
```

Comment:
So this is not addressing #219: there is still a hard cap on stdio size.

Reply:
That's true, the intention was to preserve existing behavior by making `stdioRetention = 'all'` completely identical to `keepLongStdio = true`.

I'd love to include this scenario as well and allow long output (even beyond 1MB) to be preserved when it's hit, but that may bite users who upgrade and all of a sudden find that massive files are being stored after their CI runs.

So to be clear, this PR does still address #219, just not in every possible case.

Reply:
> does still address #219, just not in every possible case

Does it?

> I should never see `...[truncated 135494 chars]...` in a failing test result […yet] I do see that even in failing test results.

which AFAICT is still true after this PR.

Still just trying to understand what this is PR is and is not changing. The new table does not show the effect on different stdio sizes, or maybe I am just not getting it. (@timja any clearer to you?)

Reply:
> Does it?

Yes, and I have personally verified this with manual testing.

> which AFAICT is still true after this PR.

If we're taking this literally and focusing on the word "never", then no, it doesn't. But like I noted before, this PR does prevent truncation from happening in some scenarios. See [here](https://github.com/C0urante/junit-plugin/blob/c34254c05e814dbe87dbe733ef41879ffe8609b8/src/main/java/hudson/tasks/junit/CaseResult.java#L231-L247).

File sizes only come into play with a [different variant](https://github.com/C0urante/junit-plugin/blob/c34254c05e814dbe87dbe733ef41879ffe8609b8/src/main/java/hudson/tasks/junit/CaseResult.java#L231-L247) of the `CaseResult::possiblyTrimStdio` method, which I've intentionally refrained from making more invasive changes to.

Reply:
So if I am using https://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html which writes `*-output.txt`, and I have a test which writes out 2Mb and output and then fails, I am still going to see the middle half truncated. Right?

Reply:
Only if the system property `maven.test.redirectTestOutputToFile`, which defaults to false, is set to true.

Reply:
As in for example https://github.com/jenkinsci/plugin-pom/blob/e4a9d6957748be5cc3e1c3026014c3db60047d7b/pom.xml#L1030-L1040.

I guess the right fix is to stop copying stdio into `junitResult.xml` in the built-in storage, instead writing/copying it to a separate file in the build directory, and then reading from that on demand in display threads. This would allow the truncation hack to be deleted without risking memory problems in the controller. What I am unsure of is whether there is then a use case for ever _not_ retaining stdio, even for passing tests. Disk space?

Reply:
> this approach potentially painting us into a corner

I do not think so, just trying to think through the implications.

Reply:
I'm going to mark this resolved as I believe I've addressed your concerns.